### DISCUSSION TITLE
Bulk update selected rows in table.  The problem is with encoded values being inserted into database

### DISCUSSION THREAD

**Discussion by bamboowonder** [Budibase/budibase#5410] [Help]
Hi there,
I have a button, with action onclick running an update query.
It sends the selected rows from table.
however the values it sends to database are encoded.  Is this a bug?  Or should I be decoding before sending?

This is a test app, once I get the functionality working, I'd like to use it for bulk updating the selected rows in a table
Here is my update query:

UPDATE "5272_list" 
SET call_results = {{callresult}}, person = 'joe' 
WHERE id = {{testcallid}}::uuid;

the {{callresult}} parameter is the problem.   I've tried both handlebars and javascript.  

handlebars:   {{ New Table.Selected rows }}
or javascript:   return $("New Table.Selected rows")

And with two rows in table selected, this is what gets updated in database:
%5B'0cb213a1-fe4c-4701-9667-208b031b420c'%5D,%5B'f0edff2c-8bec-4d45-ae39-e232fc911050'%5D

which when decoded, looks like this:
['0cb213a1-fe4c-4701-9667-208b031b420c'],['f0edff2c-8bec-4d45-ae39-e232fc911050']
which is what I want....

The id column in the table/datasource is a uuid, which is what you see above is 2 uuids for the 2 rows selected.

I tried adding a .id to the handlebars/js but that produced nothing.

I'd appreciate any pointers on what I'm doing wrong.

Thank you

**Comment by bamboowonder** [comment]
For followup:
I've modified my query to handle an array of values.  But it fails because postgres doesn't like whatever budibase is passing into $2.
Unfortunately I can't see any logs in budibase to see how it is passing the values into $2
When I manually pass this query manually, it works fine:

UPDATE "5272_teachers_list"
SET call_results = 'test4', recruiter = 'david'
WHERE id = any(ARRAY['d7a11326-c510-4dae-8bd0-f6fc4ec205d9','d7a11326-c510-4dae-8bd0-f6fc4ec205d9']::uuid[])
RETURNING *;

but whatever budibase is passing into $2 fails.

UPDATE "5272_teachers_list"
	SET call_results = $1, recruiter = 'david'
	WHERE id = any(array $2 ::uuid[])

Query Error: error: syntax error at or near "$2"

Who knows, I may wake up in the morning and realize the solution, but for now I'm stumped on this.

And, does Budibase have logs or a debugger anywhere?

**Comment by mike12345567** [comment]
Hi @bamboowonder - I see you're using a combination of datasource plus (using tables that Budibase has generated from your Postgres database) and direct queries - right now this is slightly difficult to do as Budibase encodes the primary/composite keys of your tables in the format you're seeing `%5B'0cb213a1-fe4c-4701-9667-208b031b420c'%5D` is the encoded version of a SQL table's primary key values. The way that others have handled this in the past is by decoding it in the Javascript with:
```js
return $("New Table.Selected rows").map(id => JSON.parse(decodeURIComponent(id.replace(/'/g, "\"")))[0])
```
This is admittedly quite nasty, the combination of Datasource plus and queries is something we hope to improve, although ideally in this scenario in the future we'll support bulk update and deletion directly in Budibase without the need for SQL queries to be written.

As for the second part, Budibase uses prepared statements to replace your bindings with safe SQL variables, protecting your queries with malicious user input. We are still working on improving the user experience around this, in this instance you would need to break up your selected rows into multiple, individual bindings and specify them like `any(array[{{ binding1 }}, {{binding2 }}, ...]::uuid[])`. I'll have a look at this and see if I can get Budibase to be a bit smarter around Array specifications, we already do some work for `IN` queries, so that if you specify a comma separated set of values like `WHERE column IN {{ binding }}` it'll convert this into `WHERE column IN ($1, $2, $3)` as per comma separated parts it finds in the binding - its possible we could utilise this for array definitions as well.

**Comment by bamboowonder** [comment]
Thank you Mike,

Would it better if I just manually specified direct queries for everything?  I'm totally fine with that and prefer the extra control anyway.   
Are there tradeoffs to just using direct sql everywhere instead of the datasource+ ?

Although, I just swapped out the datasource for that table to be a select * from "5272_teachers_list" instead of the datasource+ and it fell apart.   Error: Bad Gateway.     That table does have 30,000 rows, and I didn't include a LIMIT as I figured BB has that built in?

I was able to get around that error by going in and re-fetching all the tables from that connection.  Then the sql select * worked.   But then failed again.

This time, I included a limit in the raw sql query, and everything worked.  So I assume BB isn't injecting LIMIT into raw sql selects?   So would I need to build in my own pagination/page selector and bindings?

So it seemed like it might work, but I spoke too soon, when enabling "allow row selection", if I click on any row's checkbox, it selects all the rows.  Uncheck, unselects all rows.   It seems like all the checkboxes became like the "select all" checkbox at the top.   Interestingly when I click on the actual select-all checkbox, nothing changes, almost like it was disabled.

I deleted both the table and data provider and tried again from scratch.  Same behaviour.

Note, when I swap out the datasource of the dataprovider to use the table, the checkbox behaviour goes back to normal.
